hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/dao/GatewayServiceEntityDao.java

```diff
@@ -56,27 +69,31 @@
 
   @Override
   public Single<EntityResultSet> getEntities(EntityRequest request) {
-    return this.requestBuilder
-        .buildRequest(request)
-        .subscribeOn(this.boundedIoScheduler)
-        .flatMap(
-            serverRequest ->
-                this.makeRequest(request.resultSetRequest().context(), serverRequest)
-                    .flatMap(
-                        serverResponse ->
-                            baselineDao
-                                .getBaselines(
-                                    request.resultSetRequest().context(),
-                                    serverRequest,
-                                    serverResponse,
-                                    request)
-                                .flatMap(
-                                    baselineResponse ->
-                                        this.entityConverter.convert(
-                                            request, serverResponse, baselineResponse))));
+    GraphQlRequestContext context = request.resultSetRequest().context();
+    Single<EntitiesRequest> entitiesRequestSingle = this.requestBuilder.buildRequest(request);
+    return Single.zip(
+            entitiesRequestSingle,
```

Comment:
we shouldn't use the same observable/single twice. By default, assume observables/singles are unicast (they are), so each time we subscribe to it we're independently executing any underlying logic and/or server calls again. Instead, it looks like the reason for that here was to get access to both the original and transformed request in the flatMap function - my suggestion to address that would be to nest functions (any nesting solves the issue, but the functions keep it readable - pretty sure I asked for a similar change on this baseline change too). So we can do something like:

```java
@Override
  public Single<EntityResultSet> getEntities(EntityRequest request) {
    GraphQlRequestContext context = request.resultSetRequest().context();
    return this.requestBuilder
        .buildRequest(request)
        .subscribeOn(this.boundedIoScheduler)
        .flatMap(serverRequest -> this.fetchAndMapEntities(context, request, serverRequest));
  }

  private Single<EntityResultSet> fetchAndMapEntities(
      GraphQlRequestContext context, EntityRequest request, EntitiesRequest serverRequest) {
    return this.makeEntityRequest(context, serverRequest)
        .flatMap(response -> this.getEntityResultSet(request, serverRequest, response));
  }
```

Reply:
Made this suggested change